mlir/utils/performance/rocblas-benchmark-driver/rocblas-benchmark-driver.cpp

```diff
@@ -54,25 +54,28 @@ int main(int argc, char **argv) {
   rocblas_operation blasTransA, blasTransB;
   size_t lda, ldb;
 
-  // Please note: rocMLIR is using a row-major format
-  // to store matrices, while rocBLAS is using a
-  // column-major format. This means that transposition concepts are
-  // inverted
-  if (!args.transposeA) {
+  // Please note: MIGraphx and MLIR are using row-major format
+  // to store matrices, while rocBLAS is using a colum-major format.
+  // To be compliant to rocBLAS format, MIGraphx swaps the inputs
+  // and tells rocBLAS that B is nxk and A is kxm. So the result
+  // will be a nxm matrix stored in colum-major order. We can simply
+  // recover the original matrix C by reading the matrix in a row-major
+  // way.
+  if (args.transposeA) {
```

Comment:
Something feels not right here...

shouldn't lda associated with M <-> N has to change here.
I mean as I see lda is associated with args.gemmM before/after this change.

Reply:
I think the point here is about matrix C.

That matrix, when untransposed, is stored `[M, N]`. However, rocBLAS expects it to be stored `[N, M]`. This means that, our  `C = A * B` is their `tr(C) = tr(B) * tr(A)`

Reply:
That is, when we have our problem `tr?(C) = tr?(A) * tr?(B)`, that is equivalent to the rocBlas problem `tr(tr?(C)) = tr(tr?(B)) * tr(tr?(A))`, with `tr?` being either a transpose or a noop depending on the flags and `tr(tr(X))` cancelling out.

Reply:
t's really a matter of "interpretation". I will repeat what I understood to also convince myself :)

A row-major 2x3 A=[1 2 3;4 5 6] in memory is just a sequence of [1 2 3 4 5 6]. We tell rocBlas that this is a 3x2matrix and he believes us because he will read1 2 3as a column and then4 5 6as the next column. So to rocBLAS that is a3x2matrix. You do the same exact reasoning with say a3x10 Bmatrix. For a row-major library is3x10, but we tell rocBLAS that this is a 10x3matrix. Now rocBLAS gives us a10x2matrix, and it is fine for us, because we will read the first 10 consecutive elements which is a row for us. So a10x2matrixCis (in row-major) a2x10` matrix, which is exactly what we wanted.

About ld{a,b,c} , this is the number of elements between element i,j and elementi+1,j for row major and it is the number of elements between i,j and i,j+1 for column major. Going back to our example, our matrix 2x3 has lda=3 (size of a row). When we pass this to rocBLAS, it'll become a 3x2 colum-major matrix, and the number of elements between i,j and i,j+1 will still be 3 (size of a column).

Finally, transposition. For what I said above, the transposition flags are untouched. If it's not transposed for us, it won't be transposed for rocBLAS and viceversa (because we re-interpret the matrix)

Reply:
@manupak , @krzysz00  any update on this? I would like to have this in to test performance against rocblas

Reply:
That is, `blasTransA` in this branch should be `rocblas_operation_none`

Reply:
A = [1 2
 3 4
 5 6] : 3x2

B =
[a b c d
 e f g h] : 2x5

C = [1a + 2e, 1b + ...
 3a + 4e ...
 ...] : 3x4

mem(A) = [1 2 3 4 5 6]
idx_A(i, j) (2*i + j) // rocBLAS must do this 
memref<3x2xT> means that the index memref[i][j] is at (2 * i + j)

rocblas (trans off, lda = 2, m = 2, k = 3)
(i, j) -> (i + 2 * j)
(0, 0) -> 0 + (0, 1) -> 2 + (0, 2) -> 4
1? 

rocblas trans off map is (i, j) -> (i + lda * j)
rocblas trans on map is (i, j) -> (j + lda * i)

Reply:
Ok so taking A : 3x2, B : 2x4, C : 3x4 as above, the MIGraphX method is to compute

C: 4x3 = B: 4x2 * A: 2x3 , passing in the underlying buffers.

This leads to map(B) = (i + 2*j) and map(A) = (j + 2*i), which are arg-swapped versions of what MLIR uses. Similarly, C is (i, j) -> (i + 3 * j), which is also a argument-swapped version of our (i, j) -> (3 * i + j).

So, this method is equivalent to computing C = AB by tr(C) = tr(B)tr(A) implicitly.